guard/src/rules/eval_context.rs

```diff
@@ -1174,22 +1175,74 @@ impl<'value, 'loc: 'value> EvalContext<'value, 'loc> for RootScope<'value, 'loc>
     }
 }
 
-pub(crate) fn validate_number_of_params(name: &str, num_args: usize) -> Result<()> {
-    let expected_num_args = match name {
-        "join" => 2,
-        "substring" | "regex_replace" => 3,
-        "count" | "json_parse" | "to_upper" | "to_lower" | "url_decode" | "parse_string"
-        | "parse_boolean" | "parse_float" | "parse_int" | "parse_char" => 1,
-        _ => {
-            return Err(Error::ParseError(format!(
-                "no such function named {name} exists"
-            )));
+#[derive(Debug)]
+pub(crate) enum Function {
+    Join,
+    Substring,
+    RegexReplace,
+    Count,
+    JsonParse,
+    ToUpper,
+    ToLower,
+    UrlDecode,
+    ParseString,
+    ParseBoolean,
+    ParseFloat,
+    ParseInt,
+    ParseChar,
+}
+
+impl TryFrom<&str> for Function {
+    type Error = Error;
+
+    fn try_from(name: &str) -> std::result::Result<Self, Self::Error> {
+        match name {
+            "join" => Ok(Function::Join),
+            "substring" => Ok(Function::Substring),
+            "regex_replace" => Ok(Function::RegexReplace),
+            "count" => Ok(Function::Count),
+            "json_parse" => Ok(Function::JsonParse),
+            "to_upper" => Ok(Function::ToUpper),
+            "to_lower" => Ok(Function::ToLower),
+            "url_decode" => Ok(Function::UrlDecode),
+            "parse_string" => Ok(Function::ParseString),
+            "parse_boolean" => Ok(Function::ParseBoolean),
+            "parse_float" => Ok(Function::ParseFloat),
+            "parse_int" => Ok(Function::ParseInt),
+            "parse_char" => Ok(Function::ParseChar),
+            _ => Err(Error::ParseError(format!(
+                "No function with the name '{name}' exists.",
+            ))),
         }
+    }
+}
+
+pub(crate) fn validate_number_of_params(func: &Function, num_args: usize) -> Result<()> {
+    let valid_num_args = match func {
+        Function::Join => vec![2],
```

Comment:
Not sure im following why we're now returning a vector here?

Reply:
Allows for supporting optional args in the future